src/components/Terminal/setup-terminal.js

```diff
@@ -0,0 +1,98 @@
+import { Terminal as Xterm } from "xterm";
+import { FitAddon as XtermFitAddon } from "xterm-addon-fit";
+
+import XtermTerminadoAddon from "./xterm-addon-terminado";
+
+// Define jujushell API operations and codes.
+const OP_LOGIN = "login";
+const OP_START = "start";
+const CODE_OK = "ok"; // eslint-disable-line no-unused-vars
+const CODE_ERR = "error";
+const UNEXPECTED_CLOSE =
+  "Terminal connection unexpectedly closed. " +
+  "Close and re-open the terminal window to reconnect.";
+
+/**
+  Setup the terminal instance.
+  @param {String} address The address to connect the websocket to.
+  @param {Object} creds The credentials format as extracted from the bakery
+    including username, password, or the macaroons for the active user.
+  @param {Object} terminalElement The element in the DOM to render the xterm
+    instance into.
+  @returns {XTerm} The new Xterm terminal instance.
+*/
+export default function setupTerminal(address, creds, terminalElement) {
+  let terminalInstance = new Xterm({
+    theme: {
+      background: "#111"
+    }
+  });
+
+  const ws = new WebSocket(address);
+
+  ws.onopen = () => {
+    ws.send(
+      JSON.stringify({
+        operation: OP_LOGIN,
+        username: creds.user,
+        password: creds.password,
+        macaroons: creds.macaroons
+      })
+    );
+    ws.send(JSON.stringify({ operation: OP_START }));
+  };
+
+  ws.onerror = err => {
+    terminalInstance.write("Failed to open Websocket connection");
+    console.error(err);
+  };
+
+  ws.onclose = evt => {
+    // 1000 is the code for a normal closure, anything above that is abnormal.
+    if (evt && evt.code > 1000) {
+      // It is not a normal closure so we should issue an error.
+      terminalInstance.writeln(UNEXPECTED_CLOSE);
+      console.log(evt);
```

Comment:
Debugging code, or intentional?

Reply:
This one is intentional. Just adding some error logging to the console to help determine what the cause might have been.

Reply:
This one is intentional. Just adding some error logging to the console to help determine what the cause might have been.